File: HiringTracker.API/hrt.dal/src/SkillDal.js

const { SQLDal } = require('./SQLDal');
const SkillEntity = require('./entities/Skill');
const mssql = require('mssql');

class SkillDal extends SQLDal {

    init(initParams) {
        super.initDal(initParams);        
    }

    async Upsert(skills) {

        let tvp = new mssql.Table();
        tvp.columns.add('SkillID', mssql.BigInt);
        tvp.columns.add('Name', mssql.NVarChar(50));

        for(let s in skills) {
            tvp.rows.add( skills[s].SkillID, skills[s].Name );
        }

        await mssql.connect(this._config);
        let inParams = [];
        inParams['Skills']  = { type: mssql.TVP, value: tvp }; 
        
        super.execStorProcValue(mssql, 'p_Skill_Upsert', inParams);
    }

    async GetAll() {
        let skills = null;        

        await mssql.connect(this._config);

        let records = await super.execStorProcRecordset(mssql, 'p_Skill_GetAll');

        if(records) {
            skills = [];
            for(let r in records) {
                let skill = new SkillEntity();
                skill.SkillID = parseInt(records[r].SkillID);
                skill.Name = records[r].Name;

                skills.push(skill);
            }
        }

        return skills;
    }
}

module.exports = SkillDal;

File: Sources/HiringTracker.API/HRT.DAL.MSSQL/SkillDal.cs


using System;
using System.Collections.Generic;
using System.ComponentModel.Composition;
using System.Data;
using System.Data.SqlClient;
using HRT.Common;
using HRT.DAL.MSSQL;
using HRT.Interfaces;
using HRT.Interfaces.Entities;

namespace HRT.DAL.MSSQL 
{
    class SkillDalInitParams : InitParamsImpl
    {
    }

    [Export("MSSQL", typeof(ISkillDal))]
    public class SkillDal: SQLDal, ISkillDal
    {
        public IInitParams CreateInitParams()
        {
            return new SkillDalInitParams();
        }

        public void Init(IInitParams initParams)
        {
            InitDbConnection(initParams.Parameters["ConnectionString"]);
        }

        Skill Get(System.Int64? ID)
        {
            Skill result = default(Skill);

            using (SqlConnection conn = OpenConnection())
            {
                SqlCommand cmd = new SqlCommand("p_Skill_GetDetails", conn);
                cmd.CommandType = System.Data.CommandType.StoredProcedure;

                 AddParameter(   cmd, "@ID", System.Data.SqlDbType.BigInt, 0,
                                ParameterDirection.Input, false, 0, 0, string.Empty, DataRowVersion.Current, ID);
            
            
                var pFound = AddParameter(cmd, "@Found", SqlDbType.Bit, 0, ParameterDirection.Output, false, 0, 0, string.Empty, DataRowVersion.Current, 0);

                var ds = FillDataSet(cmd);

                if (ds.Tables.Count > 0 && ds.Tables[0].Rows.Count > 0)
                {
                    result = SkillFromRow(ds.Tables[0].Rows[0]);                    
                }
            }

            return result;
        }

        bool Delete(System.Int64? ID)
        {
            bool result = false;

            using (SqlConnection conn = OpenConnection())
            {
                SqlCommand cmd = new SqlCommand("p_Skill_Delete", conn);
                cmd.CommandType = System.Data.CommandType.StoredProcedure;

                            AddParameter(   cmd, "@ID", System.Data.SqlDbType.BigInt, 0,
                                ParameterDirection.Input, false, 0, 0, string.Empty, DataRowVersion.Current, ID);
            
                            var pFound = AddParameter(cmd, "@Removed", SqlDbType.Bit, 0, ParameterDirection.Output, false, 0, 0, string.Empty, DataRowVersion.Current, 0);

                cmd.ExecuteNonQuery();

                result = (bool)pFound.Value;
            }

            return result;
        }

        
        public IList<Skill> GetAll()
        {
            IList<Skill> result = base.GetAll<Skill>("p_Skill_GetAll", SkillFromRow);

            return result;
        }

        public Skill Upsert(Skill entity) 
        {
            Skill entityOut = base.Upsert<Skill>("p_Skill_Upsert", entity, AddUpsertParameters, SkillFromRow);

            return entityOut;
        }

        protected SqlCommand AddUpsertParameters(SqlCommand cmd, Skill entity)
        {
                SqlParameter pID = new SqlParameter("@ID", System.Data.SqlDbType.BigInt, 0, ParameterDirection.Input, false, 0, 0, "ID", DataRowVersion.Current, (object)entity.ID != null ? (object)entity.ID : DBNull.Value);   cmd.Parameters.Add(pID); 
                SqlParameter pName = new SqlParameter("@Name", System.Data.SqlDbType.NVarChar, 50, ParameterDirection.Input, false, 0, 0, "Name", DataRowVersion.Current, (object)entity.Name != null ? (object)entity.Name : DBNull.Value);   cmd.Parameters.Add(pName); 
        
            return cmd;
        }

        protected Skill SkillFromRow(DataRow row)
        {
            var entity = new Skill();

                    entity.ID = (System.Int64?)row["ID"];
                    entity.Name = (System.String)row["Name"];
        
            return entity;
        }
        
    }
}
